File: SIH26011/backend/app/api/buildings.py
from fastapi import APIRouter, HTTPException
from typing import List
from app.repositories.data_repository import repository
from app.schemas.pydantic_schemas import BuildingResponse, FloorResponse, PropertyUnitResponse

router = APIRouter(prefix="/api/buildings", tags=["Buildings"])

@router.get("", response_model=List[BuildingResponse])
def get_buildings():
    return [BuildingResponse(**b.__dict__) for b in repository.buildings.values()]

@router.get("/{building_id}", response_model=BuildingResponse)
def get_building_by_id(building_id: str):
    bld = repository.buildings.get(building_id)
    if not bld:
        bld = next((b for b in repository.buildings.values() if b.building_code.upper() == building_id.upper()), None)
    if not bld:
        raise HTTPException(status_code=404, detail=f"Building '{building_id}' not found.")
    return BuildingResponse(**bld.__dict__)

@router.get("/{building_id}/floors", response_model=List[FloorResponse])
def get_building_floors(building_id: str):
    floors = [f for f in repository.floors.values() if f.building_id == building_id]
    floors.sort(key=lambda x: x.floor_number)
    return [FloorResponse(**f.__dict__) for f in floors]

@router.get("/{building_id}/properties", response_model=List[PropertyUnitResponse])
def get_building_properties(building_id: str):
    units = [u for u in repository.units.values() if u.building_id == building_id]
    return [PropertyUnitResponse(**u.__dict__) for u in units]

File: backend/src/routes/buildings.js
import { Router } from "express";
import { repository } from "../repository.js";
import { findByIdOr } from "../utils.js";

const router = Router();

const getBuilding = (id) =>
  findByIdOr(repository.buildings, id, [
    "building_code",
  ]);

const notFound = (buildingId, res) =>
  res.status(404).json({
    detail: `Building '${buildingId}' not found.`,
  });

// Get all buildings
router.get("/", (_req, res) => {
  res.json(Object.values(repository.buildings));
});

// Get floors of a building
router.get(
  "/:building_id/floors",
  (req, res) => {
    const buildingId =
      req.params.building_id;

    const building =
      getBuilding(buildingId);

    if (!building) {
      return notFound(
        buildingId,
        res
      );
    }

    const floors = Object.values(
      repository.floors
    )
      .filter(
        (floor) =>
          floor.building_id ===
          building.id
      )
      .sort(
        (a, b) =>
          a.floor_number -
          b.floor_number
      );

    return res.json(floors);
  }
);

// Get properties/units of a building
router.get(
  "/:building_id/properties",
  (req, res) => {
    const buildingId =
      req.params.building_id;

    const building =
      getBuilding(buildingId);

    if (!building) {
      return notFound(
        buildingId,
        res
      );
    }

    const properties =
      Object.values(
        repository.units
      ).filter(
        (unit) =>
          unit.building_id ===
          building.id
      );

    return res.json(properties);
  }
);

// Get single building
router.get(
  "/:building_id",
  (req, res) => {
    const buildingId =
      req.params.building_id;

    const building =
      getBuilding(buildingId);

    if (!building) {
      return notFound(
        buildingId,
        res
      );
    }

    return res.json(building);
  }
);

export default router;